### PR TITLE
 sbt.internal.inc.Position: add startOffset and endOffset

### PR DESCRIPTION
**Pull request by smarter**
This is a follow-up to https://github.com/sbt/util/pull/173

### CODE REVIEW

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=1.2.x&zinc=pull/571/head&io=1.2.x&librarymanagement=1.2.x&util=1.2.x&website=1.2.x&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/936/console">is in progress...</a>

**Comment by smarter** [comment]
(Running the validation against https://github.com/sbt/util/pull/173)

**Comment by typesafe-tools** [comment]
[//]: # (Refs: 1.2.x-pull/571/head-1.2.x-1.2.x-pull/173/head-1.2.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/936/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.2.x">1.2.x               </A> | sbt/sbt@549c69463d92be021eece4f6e75883f4f692e99c
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/571">pull/571/head       </A> | sbt/zinc@6bd672bda7fa995793ce3ef98652b639a8f8009c
  io                | <A HREF="https://github.com/sbt/io/tree/1.2.x">1.2.x               </A> | sbt/io@fb0acb14a7df22f9f01bbbd0005aa0e2ae55c968
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.2.x">1.2.x               </A> | sbt/librarymanagement@f42ec6f442463d131815cefdc48ee7e1b8e00a1b
  util              | <A HREF="https://github.com/sbt/util/pull/173">pull/173/head       </A> | sbt/util@79ff000a7716cf1ca71934cc16d46bbf458558d4
  website           | <A HREF="https://github.com/sbt/website/tree/1.2.x">1.2.x               </A> | 

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.2.x&zinc=pull/571/head&io=1.2.x&librarymanagement=1.2.x&util=pull/173/head&website=1.2.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/937/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: 1.2.x-pull/571/head-1.2.x-1.2.x-pull/173/head-1.2.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/937/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.2.x">1.2.x               </A> | sbt/sbt@549c69463d92be021eece4f6e75883f4f692e99c
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/571">pull/571/head       </A> | sbt/zinc@6bd672bda7fa995793ce3ef98652b639a8f8009c
  io                | <A HREF="https://github.com/sbt/io/tree/1.2.x">1.2.x               </A> | sbt/io@fb0acb14a7df22f9f01bbbd0005aa0e2ae55c968
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.2.x">1.2.x               </A> | sbt/librarymanagement@f42ec6f442463d131815cefdc48ee7e1b8e00a1b
  util              | <A HREF="https://github.com/sbt/util/pull/173">pull/173/head       </A> | sbt/util@1b2f2cd1f1f811c672c6ec5fa040ed5365f84e28
  website           | <A HREF="https://github.com/sbt/website/tree/1.2.x">1.2.x               </A> | 

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.2.x&zinc=pull/571/head&io=1.2.x&librarymanagement=1.2.x&util=pull/173/head&website=1.2.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Review by jvican — DISMISSED** [review]
Let's not upgrade to sbt 1.2.1 here, since it causes build problems (last time I tried). Alternatively, I guess the build also fails here because we have yet to publish a jar with https://github.com/sbt/util/pull/173#pullrequestreview-145483679. Otherwise the patch LGTM!

**Comment by eed3si9n** [comment]
So should this PR target 1.x branch?

**Comment by smarter** [comment]
Any reason it couldn't get in the next 1.2 release?

**Comment by jvican** [comment]
Targeting 1.x is better, we could branch out (or cherry pick) in 1.2.2 from there.

**Comment by eed3si9n** [comment]
> Any reason it couldn't get in the next 1.2 release?

I think it depends on how critical this change is. For sbt, we don't go through RC cycle for patch releases, so we try to keep them to bug fixes. Considering that 1.3.0 will be a few months away, I'm open to a few enhancements added during 1.2.x but we need to be fairly confident that it won't cause compatibility problems.

**Comment by smarter** [comment]
It's critical to getting any BSP implementation for Scala working. My goal is for us to be able to give the Intro to FP course at EPFL that starts in mid-September using Dotty, which means using the Dotty IDE, which would really benefit from the BSP.

**Comment by jvican** [comment]
I second this change, compatibility issues with this patch are close to zero IMO.

File: internal/compiler-bridge/src/main/scala/xsbt/DelegatingReporter.scala [hunk 1]
```diff
@@ -76,18 +89,39 @@ private object DelegatingReporter {
       val line = pos.line
       val lineContent = pos.lineContent.stripLineEnd
       val offset = pos.point
-      val pointer = offset - src.lineToOffset(src.offsetToLine(offset))
+
+      // Same logic as Position#line
+      def lineOf(offset: Int) = src.offsetToLine(offset) + 1
+      def columnOf(offset: Int) = offset - src.lineToOffset(src.offsetToLine(offset))
```
Comment (smarter) [inline internal/compiler-bridge/src/main/scala/xsbt/DelegatingReporter.scala:95]: It's always been this way, but isn't it weird that lines in the zinc Reporter are 1-based and columns are 0-based ?

Comment (jvican) [reply internal/compiler-bridge/src/main/scala/xsbt/DelegatingReporter.scala:95]: Yes, I would welcome a change here, I have no idea of the rationale of the previous change.

**Review by jvican — APPROVED** [review]
LGTM

**Review by eed3si9n — APPROVED** [review]
LGTM

**Comment by retronym** [comment]
This change seems to interact poorly with Scaladoc. I think the root bug is in Scaladoc which is spitting out an invalid end position for use cases.

Minimized: https://github.com/retronym/sbt-scaladoc-aioooe

```
[info] Main Scala API documentation to /Users/jz/code/sbt-scaladoc-aioooe/target/scala-2.13/api...
[error] java.lang.ArrayIndexOutOfBoundsException: 14
[error] 	at scala.reflect.internal.util.BatchSourceFile.findLine$1(SourceFile.scala:200)
[error] 	at scala.reflect.internal.util.BatchSourceFile.offsetToLine(SourceFile.scala:203)
[error] 	at xsbt.DelegatingReporter$.lineOf$1(DelegatingReporter.scala:99)
[error] 	at xsbt.DelegatingReporter$.makePosition$1(DelegatingReporter.scala:112)
[error] 	at xsbt.DelegatingReporter$.convert(DelegatingReporter.scala:134)
[error] 	at xsbt.DelegatingReporter.info0(DelegatingReporter.scala:165)
[error] 	at scala.tools.nsc.reporters.MakeFilteringForwardingReporter.doReport(ForwardingReporter.scala:59)
[error] 	at scala.tools.nsc.reporters.FilteringReporter.info0(Reporter.scala:93)
[error] 	at scala.reflect.internal.Reporter.filteredInfo(Reporting.scala:118)
[error] 	at scala.reflect.internal.Reporter.warning(Reporting.scala:112)
[error] 	at scala.tools.nsc.doc.ScaladocAnalyzer$ScaladocTyper.$anonfun$typedDocDef$1(ScaladocAnalyzer.scala:47)
[error] 	at scala.tools.nsc.doc.ScaladocAnalyzer$ScaladocTyper.typedDocDef(ScaladocAnalyzer.scala:44)
[error] 	at scala.tools.nsc.doc.ScaladocAnalyzer$ScaladocTyper.typedDocDef$(ScaladocAnalyzer.scala:36)
```

Zinc could guard against the dodgy positions here. I'll see if I can figure out how to fix the positions in Scaladoc itself.

**Comment by retronym** [comment]
Submitted a Scala bug for this: https://github.com/scala/bug/issues/11865

**Comment by eed3si9n** [comment]
Thanks @retronym. Opened https://github.com/sbt/zinc/issues/734 for Zinc

**Pull request merged**